File: tests/distribute.test.ts
import { describe, expect, test } from "bun:test";
import { mkdirSync, mkdtempSync, readFileSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import {
	renderCommandFiles,
	renderSkillFiles,
	type SkillMeta,
	type SkillSource,
	TOOLS,
} from "../src/distribute/adapters.ts";
import { DistributeError, readConfig, writeConfig } from "../src/distribute/config.ts";
import { expectedFiles, readSkills } from "../src/distribute/generate.ts";
import { RULE_BEGIN, RULE_END, RULE_TEXT, upsertRuleBlock } from "../src/distribute/rule.ts";
import { makeSkillsSource } from "./helpers.ts";

function tempDir(): string {
	return mkdtempSync(join(tmpdir(), "specrail-distribute-"));
}

const BRAINSTORMING: SkillMeta = {
	id: "brainstorming",
	name: "specrail-brainstorming",
	description: 'Use when a design choice is open: "scope", behavior, or architecture.',
};

const SHIPPING: SkillSource = {
	id: "shipping-a-pr",
	name: "specrail-shipping-a-pr",
	description: "Use for PR lifecycle work.",
	files: [
		{ path: "SKILL.md", content: "---\nname: specrail-shipping-a-pr\n---\n\n# Shipping\n" },
		{ path: "checks.md", content: "# Checks\n" },
	],
};

describe("config 清单", () => {
	test("读写往返", () => {
		const root = tempDir();
		writeConfig(root, {
			version: "9.9.9",
			tools: ["qoder", "pi"],
			files: [".qoder/commands/specrail/brainstorming.md"],
		});
		expect(readConfig(root)).toEqual({
			version: "9.9.9",
			tools: ["qoder", "pi"],
			files: [".qoder/commands/specrail/brainstorming.md"],
		});
		expect(readFileSync(join(root, ".specrail/config.json"), "utf8").endsWith("\n")).toBe(true);
	});

	test("无清单返回 undefined", () => {
		expect(readConfig(tempDir())).toBeUndefined();
	});

	test("损坏 JSON 明确报错", () => {
		const root = tempDir();
		mkdirSync(join(root, ".specrail"), { recursive: true });
		writeFileSync(join(root, ".specrail/config.json"), "{ not json", "utf8");
		expect(() => readConfig(root)).toThrow(DistributeError);
		expect(() => readConfig(root)).toThrow(".specrail/config.json");
	});

	test("未知工具与越界路径被拒", () => {
		const root = tempDir();
		const configPath = join(root, ".specrail/config.json");
		mkdirSync(join(root, ".specrail"), { recursive: true });
		writeFileSync(
			configPath,
			JSON.stringify({ version: "1.0.0", tools: ["bogus"], files: [] }),
			"utf8",
		);
		expect(() => readConfig(root)).toThrow("tools must come from: qoder, claude, codex, pi");
		writeFileSync(
			configPath,
			JSON.stringify({ version: "1.0.0", tools: ["pi"], files: ["../escape.md"] }),
			"utf8",
		);
		expect(() => readConfig(root)).toThrow("files must be root-relative paths");
		writeFileSync(
			configPath,
			JSON.stringify({ version: "1.0.0", tools: ["pi"], files: ["/abs.md"] }),
			"utf8",
		);
		expect(() => readConfig(root)).toThrow("files must be root-relative paths");
	});
});

describe("AGENTS.md 规则块", () => {
	test("规则文本逐字固定", () => {
		expect(RULE_TEXT).toBe(
			[
				"At the start of a new piece of work, read the specrail-choosing-a-workflow skill for project onboarding or any PR lifecycle work.",
				"For other new changes, read it only when product scope, user-visible behavior, or architecture remains to decide.",
				"Continue work already routed to a workflow without routing it again; otherwise proceed directly without loading or announcing one.",
			].join("\n"),
		);
	});

	test("无文件时创建并包裹标记", () => {
		const path = join(tempDir(), "AGENTS.md");
		upsertRuleBlock(path, RULE_TEXT);
		expect(readFileSync(path, "utf8")).toBe(`${RULE_BEGIN}\n${RULE_TEXT}\n${RULE_END}\n`);
	});

	test("已有内容追加在末尾且原文完整保留", () => {
		const path = join(tempDir(), "AGENTS.md");
		const original = "# House rules\n\n- keep me\n";
		writeFileSync(path, original, "utf8");
		upsertRuleBlock(path, RULE_TEXT);
		const content = readFileSync(path, "utf8");
		expect(content.startsWith(original)).toBe(true);
		expect(content).toContain(`${RULE_BEGIN}\n${RULE_TEXT}\n${RULE_END}`);
	});

	test("二次调用幂等且旧块内文本被替换", () => {
		const path = join(tempDir(), "AGENTS.md");
		writeFileSync(path, "# House rules\n", "utf8");
		upsertRuleBlock(path, "OLD RULE");
		expect(readFileSync(path, "utf8")).toContain("OLD RULE");
		upsertRuleBlock(path, RULE_TEXT);
		const replaced = readFileSync(path, "utf8");
		expect(replaced).not.toContain("OLD RULE");
		expect(replaced.split(RULE_BEGIN)).toHaveLength(2);
		expect(replaced.startsWith("# House rules\n")).toBe(true);
		upsertRuleBlock(path, RULE_TEXT);
		expect(readFileSync(path, "utf8")).toBe(replaced);
	});
});

describe("适配器渲染", () => {
	test("工具表", () => {
		expect([...TOOLS]).toEqual(["qoder", "claude", "codex", "pi"]);
	});

	test("pi 命令：prompts 文件、description frontmatter、$@ 注入", () => {
		const [file] = renderCommandFiles("pi", [BRAINSTORMING]);
		expect(file?.path).toBe(".pi/prompts/specrail-brainstorming.md");
		expect(
			file?.content.startsWith('---\ndescription: "Use when a design choice is open: \\"scope\\"'),
		).toBe(true);
		expect(file?.content).toContain("`.pi/skills/specrail-brainstorming/SKILL.md`");
		expect(file?.content).toContain("**Input**: $@");
	});

	test("qoder 命令：specrail 命名空间的 name/description/category/tags", () => {
		const [file] = renderCommandFiles("qoder", [BRAINSTORMING]);
		expect(file?.path).toBe(".qoder/commands/specrail/brainstorming.md");
		expect(file?.content).toContain('name: "specrail-brainstorming"');
		expect(file?.content).toContain('category: "specrail"');
		expect(file?.content).toContain('tags: ["specrail"]');
		expect(file?.content).not.toContain("$@");
	});

	test("claude 命令：specrail 命名空间与 allowed-tools", () => {
		const [file] = renderCommandFiles("claude", [BRAINSTORMING]);
		expect(file?.path).toBe(".claude/commands/specrail/brainstorming.md");
		expect(file?.content).toContain('name: "specrail-brainstorming"');
		expect(file?.content).toContain("allowed-tools: Bash(specrail:*)");
		expect(file?.content).toContain("**Input**: $ARGUMENTS");
	});

	test("codex 只有技能面", () => {
		expect(renderCommandFiles("codex", [BRAINSTORMING])).toEqual([]);
		expect(renderSkillFiles("codex", [SHIPPING]).map((file) => file.path)).toEqual([
			".agents/skills/specrail-shipping-a-pr/SKILL.md",
			".agents/skills/specrail-shipping-a-pr/checks.md",
		]);
	});

	test("技能文件按工具根落位并携带兄弟文档", () => {
		const roots = {
			qoder: ".qoder/skills",
			claude: ".claude/skills",
			codex: ".agents/skills",
			pi: ".pi/skills",
		};
		for (const tool of TOOLS) {
			expect(renderSkillFiles(tool, [SHIPPING]).map((file) => file.path)).toEqual([
				`${roots[tool]}/specrail-shipping-a-pr/SKILL.md`,
				`${roots[tool]}/specrail-shipping-a-pr/checks.md`,
			]);
		}
	});
});

describe("技能源读取", () => {
	test("按名排序并解析 frontmatter", () => {
		const dir = makeSkillsSource([
			{ name: "specrail-shipping-a-pr", files: { "checks.md": "# Checks\n" } },
			{ name: "specrail-brainstorming" },
		]);
		const skills = readSkills(dir);
		expect(skills.map((skill) => skill.name)).toEqual([
			"specrail-brainstorming",
			"specrail-shipping-a-pr",
		]);
		expect(skills[0]?.id).toBe("brainstorming");
		expect(skills[0]?.description).toBe("Description of specrail-brainstorming.");
		expect(skills[1]?.files.map((file) => file.path)).toEqual(["SKILL.md", "checks.md"]);
	});

	test("忽略非技能文件与无 SKILL.md 目录", () => {
		const dir = tempDir();
		writeFileSync(join(dir, "SPEC.md"), "---\nid: skills\n---\n", "utf8");
		mkdirSync(join(dir, "not-a-skill"), { recursive: true });
		expect(readSkills(dir)).toEqual([]);
	});

	test("缺少 description 报错", () => {
		const dir = makeSkillsSource([
			{ name: "specrail-broken", skillMd: "---\nname: specrail-broken\n---\n\n# Broken\n" },
		]);
		expect(() => readSkills(dir)).toThrow("must declare a non-empty name and description");
	});

	test("目录名与 name 不一致报错", () => {
		const dir = makeSkillsSource([
			{ name: "specrail-renamed", skillMd: "---\nname: specrail-other\ndescription: x\n---\n" },
		]);
		expect(() => readSkills(dir)).toThrow('declares name "specrail-other"');
	});

	test("缺少 specrail- 前缀报错", () => {
		const dir = makeSkillsSource([
			{ name: "unprefixed", skillMd: "---\nname: unprefixed\ndescription: x\n---\n" },
		]);
		expect(() => readSkills(dir)).toThrow('must start with "specrail-"');
	});

	test("期望文件集汇总技能与命令", () => {
		const dir = makeSkillsSource([{ name: "specrail-brainstorming" }]);
		expect(expectedFiles(dir, ["qoder"]).map((file) => file.path)).toEqual([
			".qoder/commands/specrail/brainstorming.md",
			".qoder/skills/specrail-brainstorming/SKILL.md",
			".specrail/.gitignore",
		]);
		expect(expectedFiles(dir, ["codex"]).map((file) => file.path)).toEqual([
			".agents/skills/specrail-brainstorming/SKILL.md",
			".specrail/.gitignore",
		]);
		expect(expectedFiles(dir, ["pi", "codex"]).map((file) => file.path)).toEqual([
			".agents/skills/specrail-brainstorming/SKILL.md",
			".pi/prompts/specrail-brainstorming.md",
			".pi/skills/specrail-brainstorming/SKILL.md",
			".specrail/.gitignore",
		]);
	});
});

File: src/distribute/distribute_test.zig
const std = @import("std");
const adapters = @import("adapters.zig");
const config_m = @import("config.zig");
const generate = @import("generate.zig");
const rule = @import("rule.zig");

fn fixtureSkill(arena: std.mem.Allocator) !adapters.Skill {
    const files = try arena.dupe(adapters.SkillFile, &.{
        .{ .path = "SKILL.md", .content = "---\nname: specrail-demo\ndescription: \"Demo skill.\"\n---\n\n# Demo\n" },
        .{ .path = "extra.md", .content = "extra\n" },
    });
    return .{ .id = "demo", .name = "specrail-demo", .description = "Demo skill.", .files = files };
}

test "adapters render skill and command files per tool" {
    var arena_state = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena_state.deinit();
    const arena = arena_state.allocator();

    const skills = try arena.dupe(adapters.Skill, &.{try fixtureSkill(arena)});

    const qoder_commands = try adapters.renderCommandFiles(arena, .qoder, skills);
    try std.testing.expectEqual(@as(usize, 1), qoder_commands.len);
    try std.testing.expectEqualStrings(".qoder/commands/specrail/demo.md", qoder_commands[0].path);
    try std.testing.expectEqualStrings(
        "---\nname: \"specrail-demo\"\ndescription: \"Demo skill.\"\ncategory: \"specrail\"\ntags: [\"specrail\"]\n---\n\nRead and follow the skill at `.qoder/skills/specrail-demo/SKILL.md` for the current task.\n",
        qoder_commands[0].content,
    );

    const claude_commands = try adapters.renderCommandFiles(arena, .claude, skills);
    try std.testing.expectEqualStrings(".claude/commands/specrail/demo.md", claude_commands[0].path);
    try std.testing.expect(std.mem.indexOf(u8, claude_commands[0].content, "allowed-tools: Bash(specrail:*)") != null);
    try std.testing.expect(std.mem.endsWith(u8, claude_commands[0].content, "\n\n**Input**: $ARGUMENTS\n"));

    const pi_commands = try adapters.renderCommandFiles(arena, .pi, skills);
    try std.testing.expectEqualStrings(".pi/prompts/specrail-demo.md", pi_commands[0].path);
    try std.testing.expect(std.mem.endsWith(u8, pi_commands[0].content, "\n\n**Input**: $@\n"));

    const codex_commands = try adapters.renderCommandFiles(arena, .codex, skills);
    try std.testing.expectEqual(@as(usize, 0), codex_commands.len);

    const codex_skills = try adapters.renderSkillFiles(arena, .codex, skills);
    try std.testing.expectEqualStrings(".agents/skills/specrail-demo/SKILL.md", codex_skills[0].path);
    try std.testing.expectEqualStrings(".agents/skills/specrail-demo/extra.md", codex_skills[1].path);
}

test "expectedFiles sorts by path and includes the context ignore file" {
    var arena_state = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena_state.deinit();
    const arena = arena_state.allocator();

    const skills = try arena.dupe(adapters.Skill, &.{try fixtureSkill(arena)});
    const tools = try arena.dupe(adapters.Tool, &.{ .qoder, .codex });
    const files = try generate.expectedFiles(arena, skills, tools);
    var found_ignore = false;
    for (files) |file| {
        if (std.mem.eql(u8, file.path, ".specrail/.gitignore")) {
            found_ignore = true;
            try std.testing.expectEqualStrings("context/\n", file.content);
        }
    }
    try std.testing.expect(found_ignore);
    var index: usize = 1;
    while (index < files.len) : (index += 1) {
        try std.testing.expect(std.mem.order(u8, files[index - 1].path, files[index].path) == .lt);
    }
}

test "config round-trips and validates" {
    var arena_state = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena_state.deinit();
    const arena = arena_state.allocator();
    const io = std.testing.io;

    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    const root = try tmp.dir.realPathFileAlloc(io, ".", arena);

    try config_m.writeConfig(arena, io, root, .{
        .version = "0.1.0",
        .tools = &.{ .qoder, .pi },
        .files = &.{ ".specrail/.gitignore", ".qoder/commands/specrail/demo.md" },
    });
    const read_back = (try config_m.readConfig(arena, io, root)).?;
    try std.testing.expectEqualStrings("0.1.0", read_back.version);
    try std.testing.expectEqual(@as(usize, 2), read_back.tools.len);
    try std.testing.expectEqual(adapters.Tool.qoder, read_back.tools[0]);
    try std.testing.expectEqual(@as(usize, 2), read_back.files.len);

    const bad_cases = [_]struct { content: []const u8, message: []const u8 }{
        .{ .content = "[]\n", .message = ".specrail/config.json must be a JSON object" },
        .{ .content = "{}\n", .message = ".specrail/config.json must declare a specrail version" },
        .{ .content = "{\"version\":\"1\",\"tools\":[\"nope\"],\"files\":[]}\n", .message = ".specrail/config.json tools must come from: qoder, claude, codex, pi" },
        .{ .content = "{\"version\":\"1\",\"tools\":[],\"files\":[\"../x\"]}\n", .message = ".specrail/config.json files must be root-relative paths" },
    };
    const config_path = try std.fs.path.join(arena, &.{ root, config_m.CONFIG_PATH });
    for (bad_cases) |bad| {
        var file = try std.Io.Dir.cwd().createFile(io, config_path, .{ .truncate = true });
        try file.writeStreamingAll(io, bad.content);
        file.close(io);
        std.testing.expectError(error.Distribute, config_m.readConfig(arena, io, root)) catch |err| {
            std.debug.print("expected Distribute for {s}: {t}\n", .{ bad.content, err });
            return err;
        };
        try std.testing.expectEqualStrings(bad.message, @import("../errors.zig").message);
    }
}

test "rule block upserts verbatim, appends, and is idempotent" {
    var arena_state = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena_state.deinit();
    const arena = arena_state.allocator();
    const io = std.testing.io;

    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    const root = try tmp.dir.realPathFileAlloc(io, ".", arena);
    const agents = try std.fs.path.join(arena, &.{ root, "AGENTS.md" });

    try std.testing.expect(!try rule.hasRuleBlock(arena, io, agents));
    try rule.upsertRuleBlock(arena, io, agents, rule.RULE_TEXT);
    const first = try std.Io.Dir.cwd().readFileAlloc(io, agents, arena, .limited(1 << 20));
    try std.testing.expectEqualStrings(rule.RULE_BEGIN ++ "\n" ++ rule.RULE_TEXT ++ "\n" ++ rule.RULE_END ++ "\n", first);
    try std.testing.expect(try rule.hasRuleBlock(arena, io, agents));

    try rule.upsertRuleBlock(arena, io, agents, rule.RULE_TEXT);
    const second = try std.Io.Dir.cwd().readFileAlloc(io, agents, arena, .limited(1 << 20));
    try std.testing.expectEqualStrings(first, second);

    var file = try std.Io.Dir.cwd().createFile(io, agents, .{ .truncate = true });
    try file.writeStreamingAll(io, "# Notes\n\nsome prose");
    file.close(io);
    try rule.upsertRuleBlock(arena, io, agents, rule.RULE_TEXT);
    const appended = try std.Io.Dir.cwd().readFileAlloc(io, agents, arena, .limited(1 << 20));
    try std.testing.expect(std.mem.startsWith(u8, appended, "# Notes\n\nsome prose\n\n"));
    try std.testing.expect(std.mem.endsWith(u8, appended, rule.RULE_END ++ "\n"));
}

test "readSkillsFromDir validates the skill directory contract" {
    var arena_state = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena_state.deinit();
    const arena = arena_state.allocator();
    const io = std.testing.io;

    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    const root = try tmp.dir.realPathFileAlloc(io, ".", arena);

    const writeSkill = struct {
        fn call(arena2: std.mem.Allocator, io2: std.Io, base: []const u8, dir: []const u8, content: []const u8) !void {
            const dir_path = try std.fs.path.join(arena2, &.{ base, dir });
            try std.Io.Dir.cwd().createDirPath(io2, dir_path);
            const file_path = try std.fs.path.join(arena2, &.{ dir_path, "SKILL.md" });
            var file = try std.Io.Dir.cwd().createFile(io2, file_path, .{ .truncate = true });
            try file.writeStreamingAll(io2, content);
            file.close(io2);
        }
    }.call;

    try writeSkill(arena, io, root, "specrail-good", "---\nname: specrail-good\ndescription: \"Good.\"\n---\n");
    const skills = try generate.readSkillsFromDir(arena, io, root);
    try std.testing.expectEqual(@as(usize, 1), skills.len);
    try std.testing.expectEqualStrings("good", skills[0].id);
    try std.testing.expectEqualStrings("Good.", skills[0].description);

    const bad_cases = [_]struct { dir: []const u8, content: []const u8, message: []const u8 }{
        .{
            .dir = "specrail-name-mismatch",
            .content = "---\nname: specrail-other\ndescription: \"x\"\n---\n",
            .message = "specrail-name-mismatch/SKILL.md declares name \"specrail-other\"; the directory name must match it",
        },
        .{
            .dir = "no-prefix",
            .content = "---\nname: no-prefix\ndescription: \"x\"\n---\n",
            .message = "no-prefix: skill names must start with \"specrail-\"",
        },
        .{
            .dir = "specrail-no-fence",
            .content = "just prose\n",
            .message = "specrail-no-fence/SKILL.md has no YAML frontmatter",
        },
    };
    for (bad_cases, 0..) |bad, index| {
        const case_root = try std.fmt.allocPrint(arena, "{s}/case-{d}", .{ root, index });
        try writeSkill(arena, io, case_root, bad.dir, bad.content);
        std.testing.expectError(error.Distribute, generate.readSkillsFromDir(arena, io, case_root)) catch |err| return err;
        try std.testing.expectEqualStrings(bad.message, @import("../errors.zig").message);
    }
}

test "init writes expected files, then re-init with fewer tools removes the dropped ones" {
    var arena_state = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena_state.deinit();
    const arena = arena_state.allocator();
    const io = std.testing.io;

    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    const root = try tmp.dir.realPathFileAlloc(io, ".", arena);

    const skills = try arena.dupe(adapters.Skill, &.{try fixtureSkill(arena)});
    const all_tools = try arena.dupe(adapters.Tool, &.{ .qoder, .claude, .pi, .codex });

    const first = try generate.initProject(arena, io, root, all_tools, skills, "0.1.0", true);
    try std.testing.expectEqual(@as(usize, 12), first.files.len); // 1 ignore + 2 skill files x4 tools + 1 command x3 tools
    try std.testing.expect(first.rule_written);

    const user_note = try std.fs.path.join(arena, &.{ root, "notes", "mine.md" });
    try std.Io.Dir.cwd().createDirPath(io, std.fs.path.dirname(user_note).?);
    var file = try std.Io.Dir.cwd().createFile(io, user_note, .{});
    try file.writeStreamingAll(io, "user content\n");
    file.close(io);

    const fewer = try arena.dupe(adapters.Tool, &.{.qoder});
    const second = try generate.initProject(arena, io, root, fewer, skills, "0.1.0", false);
    try std.testing.expect(second.removed.len > 0);
    const claude_skill = try std.fs.path.join(arena, &.{ root, ".claude/skills/specrail-demo/SKILL.md" });
    try std.testing.expect((std.Io.Dir.cwd().statFile(io, claude_skill, .{}) catch null) == null);

    const user_content = try std.Io.Dir.cwd().readFileAlloc(io, user_note, arena, .limited(1 << 20));
    try std.testing.expectEqualStrings("user content\n", user_content);
}
